refactor: Combat Loss Arrays to Prevent Bounds Issues

jhillacre opened the pull request:
This pull request refactors the handling of `lost` and `lost_num` arrays in `obj_pnunit` to address potential bounds issues from hard-coded iterations. The updates ensure dynamic array sizing and consistent zero-indexed operations, improving stability.

- Switched to dynamic arrays to handle varying numbers of lost units safely.
- Standardized array iteration to prevent out-of-bounds errors.
- Optimized target selection by using only valid targets, enhancing performance.
- Simplified scripts and added documentation for better maintainability.

jhillacre committed:
refactor: Combat Loss Arrays to Prevent Bounds Issues

- Switched to dynamic arrays to handle varying numbers of lost units safely.
- Standardized array iteration to prevent out-of-bounds errors.
- Optimized target selection by using only valid targets, enhancing performance.
- Simplified scripts and added documentation for better maintainability.

jhillacre commented:
The scr_clean function is designed for handling player casualties, with logic and comments clearly aimed at processing damage to player units. However, it gets called with enemy unit objects in scr_shoot, causing errors since it wasn't built to handle enemy casualties.

I'll see if I can fix it or just bandaid it when I look at it tomorrow.

jhillacre committed:
fix: missed callable

jhillacre committed:
refactor: pass combat parameters as arguments instead of using object properties

- Refactored scr_clean to receive target data as arguments rather than setting them on target_object.
- Updated scr_shoot and related scripts to align with this approach.

jhillacre commented:
- Refactored `scr_clean` to receive target data as arguments rather than setting them on target_object.
- Updated `scr_shoot` and related scripts to align with this approach.

OH296 commented on scripts/scr_flavor2/scr_flavor2.gml:
```diff
@@ -316,48 +313,45 @@ function scr_flavor2(lost_units_count, target_type) {
+                    if (marine_type[him] == obj_ini.role[100][5]) {
+                        obj_ncombat.dead_jim[obj_ncombat.dead_jims] = $"A {marine_type[him]} has been critically injured!";
+                    } else {
+                        obj_ncombat.dead_jim[obj_ncombat.dead_jims] = $"{obj_ini.TTRPG[marine_co[him]][marine_id[him]].name_role()} has been critically injured!";
```
```suggestion
                        obj_ncombat.dead_jim[obj_ncombat.dead_jims] = $"{unit_struct[him].name_role()} has been critically injured!";
```

OH296 commented on scripts/scr_powers/scr_powers.gml:
```diff
@@ -512,40 +515,33 @@ function scr_powers(power_set, power_count, enemy_target, unit_id) {
+            }
+
+            // Update unit counts
+            if (unit.IsSpecialist("dreadnoughts")) {
```
```suggestion
            var armour_data = unit.armour_data();
            var is_dread = false;
            if (is_struct(armour_data)){
                 is_dread = armour_data.has_tag("dreadnought);
            }
            if (is_dread) {
```
This will be more adaptable moving forward, but worth adding a //TODO to incorporate logic into IsSpecialist method

OH296 commented on scripts/scr_shoot/scr_shoot.gml:
```diff
@@ -71,17 +75,16 @@ function scr_shoot(weapon_index_position, target_object, target_type, damage_dat
-				target_object.hostile_shots = hit_number;
+				hostile_shots = hit_number;
 				if (wep_owner[weapon_index_position] = "assorted") then target_object.hostile_shooters = 999;
 				if (wep_owner[weapon_index_position] != "assorted") then target_object.hostile_shooters = 1;
```
```suggestion
				if (wep_owner[weapon_index_position] == "assorted"){ target_object.hostile_shooters = 999;}
				else if (wep_owner[weapon_index_position] != "assorted"){ target_object.hostile_shooters = 1;}
```

optomise with else if statement

OH296 reviewed (changes requested):
noticed a few things to add

OH296 commented on objects/obj_pnunit/Create_0.gml:
```diff
@@ -93,23 +93,20 @@ repeat(71){i+=1;
-}
+// These arrays are the losses on any one frame.
+// Let them resize as required.
+// Hardcoded lengths lead to bounds issues when hardcoded values disagree.
```
any idea how many hardcoded values does this leave remaining? I managed to remove most of the base unit data ones, combined with this thats a large amount of the infrastructure of combat rewritten. I wander we shouldn't work on some batch testing methods for combat before this and my previous rework goes live in a release?

jhillacre replied:
> any idea how many hard-coded values does this leave remaining?

I'd suspect anywhere there is a repeat right now for:
- hard-coded array lengths that may not match
- off by one array index management, either treating 1 as first or missing last.

I didn't want to pull too many more threads before the sweater is a ball of yarn on this PR.

OH296 replied:
> > any idea how many hard-coded values does this leave remaining?
> 
> I'd suspect anywhere there is a repeat right now for:
> 
> * hard-coded array lengths that may not match
> * off by one array index management, either treating 1 as first or missing last.
> 
> I didn't want to pull too many more threads before the sweater is a ball of yarn on this PR.

fully makes sense big pr's have a habit of making this code base go wonky

OH296 commented on objects/obj_pnunit/Step_0.gml:
```diff
@@ -1,4 +1,9 @@
+
 highlight=0;
 var diff,siz,pos;
 diff=0;pos=880;siz=min(400,(men*0.5)+(veh*2.5)+(dreads*2));
```
```suggestion
var diff=0;
var pos=880;
var siz=min(400,(men*0.5)+(veh*2.5)+(dreads*2));
```

OH296 commented on scripts/scr_clean/scr_clean.gml:
```diff
@@ -1,225 +1,187 @@
+                }
+
+                // Select a random vehicle from the valid list
+                var random_index = irandom(array_length(valid_vehicles) - 1);
```
```suggestion
                var random_index = array_random(valid_vehicles);
```
i've made up a few functions to simplify things like this

OH296 commented on scripts/scr_player_combat_weapon_stacks/scr_player_combat_weapon_stacks.gml:
```diff
@@ -128,14 +122,17 @@ function scr_player_combat_weapon_stacks(){
-	                dreaded=true;
-	            }
-	            //if (marine_mobi[g]="Bike") then scr_special_weapon("Twin Linked Bolters",g,true);
+                var is_dreadnought = unit.armour()=="Dreadnought";
```
```suggestion
            var armour_data = unit.armour_data();
            var is_dreadnought = false;
            if (is_struct(armour_data)){
                 is_dreadnought = armour_data.has_tag("dreadnought);
            }
```
more adaptable in case we bring in variant items

OH296 commented on scripts/scr_powers/scr_powers.gml:
```diff
@@ -512,40 +515,33 @@ function scr_powers(power_set, power_count, enemy_target, unit_id) {
-        
-	        if (unit.IsSpecialist("dreadnoughts")) then dreads-=1;
-	        if (obj_ncombat.red_thirst=1) and (marine_type[unit_id]!="Death Company") then obj_ncombat.red_thirst=2;
+        if (unit.hp() < 0){
```
```suggestion
        if (unit.hp() < 0){//TODO create is_dead function to remove repeats of this log
```

OH296 reviewed (changes requested):
noticed a few further things made as suggestions so shouldn't be any burden

jhillacre committed:
refactor: improve code clarity and consistency in unit handling

- Simplified conditional logic for weapon ownership checks
- Standardized dreadnought detection using armour_data and tags
- Added TODO for creating is_dead function to reduce repetition
- Replaced random index selection with our array_random utility
- Split variable declarations for better readability

Co-authored-by: Nelsonh <81228864+OH296@users.noreply.github.com>

jhillacre commented:
@OH296 thanks for the review & suggestions.

`armour_data` doesn't appear to be on my branch at the moment, but `get_armour_data` does. Is [EttyKitty:compile/main](https://github.com/EttyKitty/ChapterMaster/tree/compile/main) what a PR should be based on?

EttyKitty commented:
> `armour_data` doesn't appear to be on my branch at the moment, but `get_armour_data` does.

I assume that's just a typo. he probably meant get_armour_data().
armour_data doesn't seem to exist even on the other branches.


> Is [EttyKitty:compile/main](https://github.com/EttyKitty/ChapterMaster/tree/compile/main?rgh-link-date=2024-10-20T14%3A57%3A05Z) what a PR should be based on?

probably, yeah. the other active branch is for draw_unit_image changes, so probably unrelated to this pr.

jhillacre committed:
refactor: clean up array access and formatting inconsistencies

- Simplified array_random usage in scr_clean by directly using the returned element
- Removed redundant checks for undefined 'lost' before updating losses
- Fixed inconsistent indentation in scr_player_combat_weapon_stacks
- Corrected typo in scr_powers: renamed armour_data to get_armour_data and fixed string quotation

jhillacre committed:
docs: finish scr_clean args documentation.

jhillacre committed:
chore: partial whitespace cleanup.

- removed mixed tab & space indentation for just spaces.
- removed trailing spaces.

jhillacre commented:
I was able to get in-game, and deal with orks, tyriands and traitors around a starting area in both raids, attacks, and garrison defence, with vehicles on the field.

I think this is ready.

OH296 commented:
> @OH296 thanks for the review & suggestions.
> 
> `armour_data` doesn't appear to be on my branch at the moment, but `get_armour_data` does. Is [EttyKitty:compile/main](https://github.com/EttyKitty/ChapterMaster/tree/compile/main) what a PR should be based on?

yes is as etty says so apologies for this, i was just working from github and didn't cross check across the code base

OH296 reviewed (approved):
aight gave a few tests and it seems to run fine, should be good @EttyKitty  i see no reason why this shouldn't go in the next compile which i'm thinking is going to be 0.9.4.0? it has quite a lot more going in it than a standard fix

jhillacre committed:
fix: missed translating valid index to veh index.
